Add MIP Reconstruction in trigger

cbell1402 opened the pull request:
I am updating _ldmx-sw_, here are the details.

### What are the issues that this addresses?
This resolves #1758.

I'm adding ECAL MIP reconstruction with @therwig's addition of HCAL MIP reconstruction. 

For ECAL MIPs, we record the length of the track, the number of hits in the track, the number of holes (length - hits), and the sum of energy in a region around the track (isolation energy). The goal with ECAL MIP tracking is to reconstruct isolated tracks with few holes in the track. To achieve this, we have two handles `isolationECut` and `hole_fraction`.  First, `isolationECut` ensures that the sum of energy in a radius (currently 50mm) around a hit in the MIP track does not exceed some value (currently 180 MeV). Second, `hole_fraction` ensures that there are not greater than some number of holes (currently 20%) in the track. Both of these handles ensure that reconstruction keeps a good amount of the signal (MIPs) while not triggering too much on background (electrons). The background events that do pass reconstruction typically leave short (<15 layers) tracks, whereas muons leave longer tracks (>20 layers). Please see the plots below. 

For HCAL MIPs, we record the length of the track, the number of hits in the track, the number of holes (length - hits). The goal here is to reconstruct straight MIP tracks in the back HCAL. There are **NOT** `isolationECut` and `hole_fraction` handles here but the algorithm performs well for signal while only reconstructing very few, very short track background events. Please see the plots below.

This PR also adds the sum of energy in the side HCAL to the trigger ntuples. 

### Flow chart for the logic of the ECAL MIP tracking algorithm 
![ECAL-Flow-Chart](https://github.com/user-attachments/assets/8b59e7cb-f932-437d-88cc-e0eaa9896a95)
## Plots
Below is reconstructed MIP length in the ECAL produced from an electron beam. This is from 100k events. Most entries are on the short length side with less than 20 entries greater than 15 layers in length.

![ele_100k_logy_length](https://github.com/user-attachments/assets/c458d9a0-7d19-438c-812e-59894a8def3d)

Next is reconstructed MIP length in the ECAL produced from 10k γ to μμ events.

![gammamumu_10k_logy_length](https://github.com/user-attachments/assets/d781cc70-8b65-48d0-85b3-a9cb770eedf9)

Finally, for the ECAL, is reconstructed MIP length for 10k events of linear energy (0-1GeV) muon gun. 

![muon_10k_logy_length](https://github.com/user-attachments/assets/25cb28c8-4215-421c-9ca6-a77e8ea0a207)

Below is reconstructed MIP length in the HCAL produced from an electron beam. This is from 100k events.

![ele_100k_logy_length_hcal](https://github.com/user-attachments/assets/42f5143b-babf-43e1-903a-bc3bc6e07d1f)

Next is reconstructed MIP length in the HCAL produced from 10k γ to μμ events.

![gammamumu_10k_logy_length_hcal](https://github.com/user-attachments/assets/ebaf1066-c269-4681-8902-b4ede70274be)

Finally, for the HCAL, is reconstructed MIP length for 10k events of linear energy (0-1GeV) muon gun. 

![muon_10k_logy_length_hcal](https://github.com/user-attachments/assets/ef1c3e0e-22d3-4cb5-be62-914f645c18f0)

## Check List
- [x] I successfully compiled _ldmx-sw_ with my developments.
- [x] I read, understood and follow the [coding rules](https://ldmx-software.github.io/developing/coding-rules.html).
- [x] I ran my developments and the following shows that they are successful.
<!-- put plots or some other proof that your developments work and do the intended function -->

Commit:
add trig Hcal hits

Commit:
add first pass of mip track reco

Commit:
add mip class

Commit:
add basic mip analyzer

cbell1402 committed:
begin ecal mip tracking

cbell1402 committed:
add ecal mip via dR and energy req

cbell1402 committed:
allow seeds from any layer, loosen dR

cbell1402 committed:
correct mip tracks such that only max length is recorded

cbell1402 committed:
account for layer shift, min/max energy, usedhits

cbell1402 committed:
begin adding side hcal sums

cbell1402 committed:
add v15 to v14.detectors.valid

cbell1402 committed:
side hcal sums correctly now

cbell1402 committed:
minor changes with spacing

cbell1402 committed:
add isolation requirment for tracks

cbell1402 committed:
beginning to only consider isolated mip tracks

cbell1402 committed:
move from energy prefilter to in-line filter

cbell1402 committed:
add hole fraction, ntuple objects

cbell1402 committed:
add hcal mips to ntuple

cbell1402 committed:
cleanup debug comments

cbell1402 committed:
add growing search window if holes exist

cbell1402 committed:
adjust isolation energy to capture more di-muon events

cbell1402 committed:
ran just format-cpp

cbell1402 committed:
Merge branch 'trunk' into add_ecalmiptracking

cbell1402 committed:
add pass name to exist checks and correct DECLARE_PRODUCER

tvami commented on DetDescr/python/HcalGeometry.py:
```diff
@@ -487,5 +487,5 @@ def make_v14(self):
         # added the reduced geometry temporarily, for the final geometry
         # we should have a new function "reduced()" with the prototype geom
-        self.v14.detectors_valid = ["ldmx-det-v14", "ldmx-det-v14.*", "ldmx-reduced", "ldmx-reduced-v2","ldmx-lyso-r1-v14", "ldmx-lyso-r1-v14.*", "ldmx-det-v15","ldmx-det-v15.*"]
+        self.v14.detectors_valid = ["ldmx-det-v14", "ldmx-det-v14.*", "ldmx-reduced", "ldmx-reduced-v2","ldmx-lyso-r1-v14", "ldmx-lyso-r1-v14.*", "ldmx-det-v15", "ldmx-det-v15.*"]
```
Please dont make this change, so https://github.com/LDMX-Software/ldmx-sw/pull/1754 doesnt have to be rebased

tvami commented on Trigger/Algo/include/Trigger/Event/TrigCaloHit.h:
```diff
@@ -37,18 +40,30 @@ class TrigCaloHit {
+  float x() const { return x_; }
+  float y() const { return y_; }
+  float z() const { return z_; }
+  float e() const { return e_; }
```
I know this wasnt you, but since you are touching this, these should be renamed to not be single letter variables, see I.1 in https://ldmx-software.github.io/developing/coding-rules.html

tvami commented on Trigger/Algo/include/Trigger/Event/TrigCaloHit.h:
```diff
@@ -37,18 +40,30 @@ class TrigCaloHit {
+  void setLayer(int x) { layer_ = x; }
+  void setStrip(int x) { strip_ = x; }
+  void setModule(int x) { modulesection_ = x; }
+  void setSection(int x) { modulesection_ = x; }
```
while this is correct, it's also pretty confusing, I suggest 
```
void setSection(int x) { modulesection_ = x; }
```
to become
```
void setSection(int modulesection) { modulesection_ = modulesection; }
```

tvami commented on Trigger/Algo/include/Trigger/Event/TrigCaloHit.h:
```diff
@@ -37,18 +40,30 @@ class TrigCaloHit {
+  int modulesection_{0};
 
   /// ROOT Dictionary class definition macro
   ClassDef(TrigCaloHit, 1);
```
```suggestion
  ClassDef(TrigCaloHit, 2);
```

tvami commented on Trigger/Algo/include/Trigger/Event/TrigMip.h:
```diff
@@ -0,0 +1,56 @@
+ public:
+  TrigMip() = default;
+
+  // TrigMip(float x, float y, float z, float e = 0);
```
```suggestion
```

tvami commented on Trigger/Algo/include/Trigger/Event/TrigMip.h:
```diff
@@ -0,0 +1,56 @@
+
+  bool operator<(const TrigMip &h) { return length_ < h.length_; }
+
+  // void Clear();
```
```suggestion
```

tvami commented on Trigger/Algo/include/Trigger/Event/TrigMip.h:
```diff
@@ -0,0 +1,56 @@
+  int nHits_{0};
+  int nHoles_{0};
+  int length_{0};
+  float SumEinIsolationRegion_{0.0f};
```
rename according go I.3 https://ldmx-software.github.io/developing/coding-rules.html

tvami commented on Trigger/Algo/include/Trigger/HcalTPSelector.h:
```diff
@@ -0,0 +1,53 @@
+ */
+
+#ifndef HCALTPSELECTOR_H
+#define HCALTPSELECTOR_H
```
```suggestion
#ifndef TRIGGER_HCALTPSELECTOR_H
#define TRIGGER_HCALTPSELECTOR_H
```

tvami commented on Trigger/Algo/include/Trigger/HcalTPSelector.h:
```diff
@@ -0,0 +1,53 @@
+
+// LDMX Framework
+// #include "DetDescr/HcalGeometry.h"
+// #include "Hcal/HcalTriggerGeometry.h"
```
```suggestion
```

tvami commented on Trigger/Algo/include/Trigger/HcalTPSelector.h:
```diff
@@ -0,0 +1,53 @@
+  // name of collection for HcalTPs to be passed as input
+  std::string combinedQuadCollName_;
+  // name of output collection
+  std::string passCollName_;
```
I.3

tvami commented on Trigger/Algo/include/Trigger/HcalTPSelector.h:
```diff
@@ -0,0 +1,53 @@
+  std::string passCollName_;
+
+  // unsigned int maxCentralTPs_{12};
+  // unsigned int maxOuterTPs_{8};
```
del

tvami commented on Trigger/Algo/include/Trigger/TrigMipReco.h:
```diff
@@ -0,0 +1,46 @@
+ */
+
+#ifndef HCALTPSELECTOR_H
+#define HCALTPSELECTOR_H
```
```suggestion
#ifndef TRIGGER_TRIGMIPRECO_H
#define TRIGGER_TRIGMIPRECO_H
```
was a copy paste error here originally too

tvami commented on Trigger/Algo/include/Trigger/TrigMipReco.h:
```diff
@@ -0,0 +1,46 @@
+  bool calorimeterTypeIsHcal_;
+
+  float minEnergy_;  // set in configure()
+  float maxEnergy_;  // set in configure()
```
I.3

tvami commented on Trigger/Algo/python/trigger_energy_sums.py:
```diff
@@ -29,6 +37,24 @@ def __init__(self, instance_name = 'myTrigHcalEnergySum') :
+    def __init__(self, instance_name = 'myHCalTrigMipReco') :
+        super().__init__(instance_name , 'trigger::TrigMipReco','Trigger')
+        self.hitCollName = "hcalTrigHits"
+        self.passCollName = "hcalTrigMIPs"
```
why the passname not just be "" in default?

tvami commented on Trigger/Algo/src/Trigger/Event/TrigMip.cxx:
```diff
@@ -0,0 +1,17 @@
+  //   nHits_ = 0;
+  //   nHoles_ = 0;
+  //   length_ = 0;
+  // }
```
You should either delete this, or put it back

tvami commented on Trigger/Algo/src/Trigger/HcalTPSelector.cxx:
```diff
@@ -0,0 +1,39 @@
+    ldmx::HcalTriggerID combo_id(tp.getId());
+
+    int adc = tp.getPrimitive();
+    double e = adc * 1.2 / 72.961;  // ADC to MeV
```
I.1, also where is the 1.2 and 72.961 coming from? See III.13

cbell1402 replied:
It looks like these values come from [here ](https://github.com/LDMX-Software/ldmx-sw/blob/3283af5fa8f873bce2c7a84eddec280e038b6f8f/Trigger/Algo/src/Trigger/TrigHcalEnergySum.cxx#L21). I will add in some comments to clarify what those numbers mean.

tvami commented on Trigger/Algo/src/Trigger/TrigMipReco.cxx:
```diff
@@ -0,0 +1,252 @@
+#include "Trigger/TrigMipReco.h"
+
```
please revise this for I.3, and I guess I.2 too

tvami commented on Trigger/Algo/src/Trigger/TrigMipReco.cxx:
```diff
@@ -0,0 +1,252 @@
+
+    // straight MIP reco
+    TrigMipCollection mips;
+    for (int i = 0; i < 5; i++) {
```
why 5?

cbell1402 replied:
I think it is because there are only five elements in the even/odd start matrices. I will add something to clarify that.

tvami commented on Trigger/Algo/src/Trigger/TrigMipReco.cxx:
```diff
@@ -0,0 +1,252 @@
+    // straight MIP reco
+    TrigMipCollection mips;
+    for (int i = 0; i < 5; i++) {
+      if (oddStart[i] < 80) {  // start in first 5 layers
```
make 80 configurable, or a const expression

tvami commented on Trigger/Algo/src/Trigger/TrigMipReco.cxx:
```diff
@@ -0,0 +1,252 @@
+  } else {                 // ECAL MIP Reconstruction
+    float radiusCut = 50;  // mm
+    int maxLayer = 32;
+    int minTrackLength = 5;  // Adjustable
```
all these should be configurable from the python

tvami commented on Trigger/Algo/include/Trigger/Event/TrigMip.h:
```diff
@@ -0,0 +1,56 @@
+};
+}  // namespace trigger
+
+#endif  // TRIGGER_EVENT_TRIGCALOCLUSTER_H
```
fix the comment here too

tvami commented on Trigger/Algo/src/Trigger/TrigMipReco.cxx:
```diff
@@ -0,0 +1,252 @@
+        m.setEndLayer(oddEnd[i]);
+        m.setNHits(oddCounts[i]);
+        m.setLength(oddEnd[i] - oddStart[i] + 1);
+        m.setNHoles(m.length() / 2 - m.nHits());
```
can `m.length() / 2 - m.nHits()` ever be negative?

cbell1402 replied:
Yes, it could be. I will correct that.

tvami commented on Trigger/Algo/include/Trigger/Event/TrigMip.h:
```diff
@@ -0,0 +1,56 @@
+
+  virtual ~TrigMip() = default;
+
+  bool operator<(const TrigMip &h) { return length_ < h.length_; }
```
```suggestion
  bool operator<(const TrigMip &h) const { return length_ < h.length_; }
```

tvami reviewed (changes requested):
* Please revise the coding rules, and apply them wherever you add new code.
 * you should add the new producers in the target GENIE CI file so they are being tested
 * I'd like to see some timing analysis too, how much extra time it takes to run this MIP tracking, please note we just removed MIP tracking in the ECAL veto to save time, it will be funny if this takes longer than that and we add it back at an earlier level

cbell1402 committed:
add tvami suggestions, mostly following code guidelines better

cbell1402 committed:
add additional tvami suggestions, reformat TrigMipReco files

cbell1402 committed:
remove comments in trigmipreco...

cbell1402 committed:
add time logging and add new producers to GENIE CI

cbell1402 commented:
Okay, I've added timing logging and added the new producers to the GENIE CI file. 
The timing returns:
```
[ myHCalTrigMipReco ] 100  info: AVG Time/Event: 0.018 ms
[ myECalTrigMipReco ] 100  info: AVG Time/Event: 0.025 ms
```
I'm not sure how that compares to MIP tracking in the ECAL veto.

tvami reviewed (approved):
Looks pretty good! Thanks Charles!

tomeichlersmith commented on Trigger/Algo/include/Trigger/Event/TrigMip.h:
```diff
@@ -0,0 +1,54 @@
+#include "TObject.h"  //For ClassDef
+
+namespace trigger {
+
+// Forward declaration needed by typedef
+class TrigMip;
+typedef std::vector<TrigMip> TrigMipCollection;
```
I don't know who started this trend, but it is not needed and (I would argue) more confusing than just using `std::vector<TrigMip>` in your producers/analyzers.

```suggestion
```

This more of an opinion though so you can ignore it if need be.

cbell1402 replied:
I also corrected this in `/Trigger/Algo/src/Trigger/EcalTPSelector.cxx`

tomeichlersmith commented on Trigger/Algo/include/Trigger/TrigMipReco.h:
```diff
@@ -0,0 +1,72 @@
+  // calorimeter_type_is_hcal_ boolean
+  bool calorimeter_type_is_hcal_;
+
+  int max_layer_;
+  int min_track_length_;
+  float hcal_min_energy_;    // MIP peak is 10-11 MeV
+  float ecal_min_energy_;    // MIP peak around 17 MeV
+  float ecal_max_energy_;    // MeV
+  float search_radius_;      // mm
+  float isolation_e_cut_;    // MeV; Change as needed
+  float hole_fraction_max_;  // No more than 20% hole fraction ; Change
+                             // as needed ; Converts tracks to MIP objects
```
Please use doxygen comments above the variable.
If your comment is only one line, you can use `///` instead of the multi-line `/** ... */`.

e.g.
```cpp
  /// MIP peak is 10-11 MeV usually
  float hcal_min_energy_;
```

tomeichlersmith commented on Trigger/Algo/src/Trigger/Event/TrigCaloHit.cxx:
```diff
@@ -3,7 +3,15 @@
 ClassImp(trigger::TrigCaloHit)
```
```suggestion
ClassImp(trigger::TrigCaloHit);
```
adding a unnecessary semicolon after this macro helps the formatter not get confused.

tomeichlersmith commented on Trigger/Algo/src/Trigger/Event/TrigMip.cxx:
```diff
@@ -0,0 +1,14 @@
+#include "Trigger/Event/TrigMip.h"
+
+ClassImp(trigger::TrigMip)
```
```suggestion
ClassImp(trigger::TrigMip);
```

tomeichlersmith commented on Trigger/Algo/src/Trigger/TrigMipReco.cxx:
```diff
@@ -0,0 +1,300 @@
+  max_layer_ = ps.getParameter<int>("max_layer", 32);
+  search_radius_ = ps.getParameter<float>("search_radius", 50.0f);  // mm
+  min_track_length_ = ps.getParameter<int>("min_track_length", 5);
+  isolation_e_cut_ = ps.getParameter<float>("isolation_e_cut",
+                                            180.0f);  // MeV; Change as needed
```
Our format is forcing the lines to be 80 characters long. On lines like these, it is helpful to put the comment above the line so that the code can look prettier.

More of an aesthetic choice though so you can ignore this if you don't have the time.

tomeichlersmith commented on Trigger/Algo/src/Trigger/TrigMipReco.cxx:
```diff
@@ -0,0 +1,300 @@
+    }
+
+    // straight MIP reco
+    TrigMipCollection mips;
```
see comment from earlier about this just being a `std::vector`

tomeichlersmith reviewed (commented):
This is looking great :) I mainly have aesthic/cleanup comments.

As for the timing, this algorithm is much faster than the EcalVeto one.
```
[ ecalVeto ] 10000  info: mip_tracking_setup     Avg Time/Event = 0.228 ms
[ ecalVeto ] 10000  info: set_variables           Avg Time/Event = 0.013 ms
[ ecalVeto ] 10000  info: bdt_variables           Avg Time/Event = 0.161 ms
[ ecalMipTracking ] 10000  info: AVG Time/Event: 0.39 ms
[ ecalMipTracking ] 10000  info: Breakdown::
[ ecalMipTracking ] 10000  info: straight_tracks        Avg Time/Event = 0.362 ms
[ ecalMipTracking ] 10000  info: linreg_tracks        Avg Time/Event = 0.023 ms
```
However, I'm unsure on if that's because of the different types of events. The EcalVeto one is using Ecal PN events which tend to have a lot of hits from an electron and/or photon shower.

cbell1402 committed:
add tomeichlersmith suggestions

tvami commented on Trigger/Algo/include/Trigger/Event/TrigCaloHit.h:
```diff
@@ -7,8 +7,8 @@
 // Forward declaration needed by typedef
-class TrigCaloHit;
-typedef std::vector<TrigCaloHit> TrigCaloHitCollection;
+// class TrigCaloHit;
+// typedef std::vector<TrigCaloHit> TrigCaloHitCollection;
```
```suggestion
```

tvami commented on Trigger/Algo/include/Trigger/Event/TrigCaloHit.h:
```diff
@@ -7,8 +7,8 @@
 namespace trigger {
 
 // Forward declaration needed by typedef
```
```suggestion
```

cbell1402 committed:
remove leftover comments

cbell1402 commented:
I just finished your last change @tvami. Thanks to you and @tomeichlersmith for all of the help! Once the tests pass, am I okay to press "Squash and merge"?

tvami commented:
Yes, when https://github.com/LDMX-Software/ldmx-sw/actions/runs/16352747766/job/46203472038?pr=1764 finishes

tvami commented:
@cbell1402 you can merge now, please have the squash msg nice and descriptive
